### PR TITLE
tests: enable bundle functional tests for v2 compiled pipelines

### PR DESCRIPTION
**Pull request by DnPlas**
The recent upgrades in kubeflow pipelines introduced a new version of the SDK that is compatible with the newer backend (pipelines 2.0.3). This commit ensures that the CI tests both a pipeline v1 and v2 to ensure good coverage.

According to the [Version Compatibility](https://www.kubeflow.org/docs/components/pipelines/v2/version-compatibility/) table, pipelines that are compiled using the 1.8 SDK are still compatible with the 2.0 backend, hence the need to cover both cases in the CI for our solution. 
From the official documentation (and from experimentation) we can also tell that to be able to create runs from a compiled v1 and v2 pipelines using the Python SDK, their corresponding SDK versions must be used; that is, creating a v2 pipeline run can only be done with the v2 SDK, and the same applies for v1 compiled pipelines.

> NOTE: this PR proposes a change in the deletion method for an existing pipeline. The new backend expects the pipeline version to be deleted before the actual pipeline can be deleted; otherwise the backend will complain with the following msg:

```
HTTP response headers: HTTPHeaderDict({'X-Powered-By': 'Express', 'content-type': 'application/json', 'date': 'Thu, 16 Nov 2023 15:46:04 GMT', 'content-length': '799', 'connection': 'close'})
HTTP response body: {"error":"Failed to delete pipeline (v1beta1) c26ba6b0-f259-469a-836a-178e98d1de2c. Check error stack: Invalid input error: Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)","code":3,"message":"Failed to delete pipeline (v1beta1) c26ba6b0-f259-469a-836a-178e98d1de2c. Check error stack: Invalid input error: Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)","details":[{"@type":"type.googleapis.com/google.rpc.Status","code":3,"message":"Failed to delete pipeline with id c26ba6b0-f259-469a-836a-178e98d1de2c as it has existing pipeline versions (e.g. 869951bf-9299-47a2-8d35-4d2038bcb5f6)"}]}
```

These changes are noted [here](https://github.com/canonical/kfp-operators/pull/383/files#diff-cf2552fea155df80690c03878d4970941f12f251449d93388c1121ca4fcd34e2R45), for example.

#### Testing environment
* microk8s 1.25-strict/stable
  * ensure the following addons are enabled `dns hostpath-storage rbac metallb:10.64.140.43-10.64.140.49`
* juju 3.1/stable, a juju controller already bootstrapped is also required
* juju model `kubeflow`
* tox installed

Testing v1 pipelines
Execute the functional tests with: `tox -e bundle-integration-v1 -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2`
The above will build and deploy all kfp-operators charms from [dnplas-add-kfp-v2-integration-tests](https://github.com/canonical/kfp-operators/tree/dnplas-add-kfp-v2-integration-tests) and then execute the following test cases:
* upload v1, v2 pipeline
* create a run with a v1, v2 pipeline
* create a recurring run with a v2 pipeline
* apply a viewer
* check the viz server
 
Testing v2 pipelines
Same steps as above, but the command is `tox -e bundle-integration-v2 -- --model kubeflow --bundle=./tests/integration/bundles/kfp_latest_edge.yaml.j2`

> NOTE: bundle functional tests are expected to pass in this PR.

TODO:

- [x] provide testing instructions using tox

### CODE REVIEW

File: tests/integration/test_kfp_functional.py
Comment (ca-scribner) [inline tests/integration/test_kfp_functional.py:112]: not a requirement, but I think improves clarity
```suggestion
        run_name=f"test-run-sdk-{kfp_version}",
```

File: tests/integration/pipelines/pipeline_container_no_input.yaml
Comment (ca-scribner) [inline tests/integration/pipelines/pipeline_container_no_input.yaml:2]: I think this an artifact generated by running `tests/integration/pipelines/pipeline_container_no_input.py`?  Can we add a note here saying this is a generated file and it should not be edited manually?
```suggestion
# PIPELINE DEFINITION
# Name: v2-container-component-no-input
# Note: This file is auto-generated by running `pipeline_container_no_input.py`.  
#           Do not edit this file manually, edit and run the `.py`
```

File: requirements-integration.in
Comment (ca-scribner) [inline requirements-integration.in:8]: Do we need `kfp` here at all?  I think this is only needed when trying to update the python examples (eg: whenever a developer is running `tests/integration/pipelines/*.py`).  Can't we just remove it entirely, and note which version is needed in the `*.py` files?  Maybe by providing a requirements.txt for each example `.py`.

Since there are examples for both sdk v1 and v2, no single version of kfp applies to all tests here.

**Review by ca-scribner — COMMENTED** [review]
generally lgtm, but added a few comments.

Should the bundle tests be passing after this PR?

File: requirements-integration.in
Comment (DnPlas) [reply requirements-integration.in:8]: Unfortunately we need kfp. We use it to create a client to talk to the `apiserver` when we upload pipelines, create runs, experiments, etc.

Comment (ca-scribner) [reply requirements-integration.in:8]: ohhh right like [here](https://github.com/canonical/kfp-operators/blob/240adae542544b35194e86e18715dfa601c1bc5e/tests/integration/conftest.py#L80).  Ok makes sense

**Review by ca-scribner — DISMISSED** [review]
lgtm.  I wasn't sure if integration tests were supposed to pass here or if they were being resolved in one of the other parallel PRs.  Worst case, this is resolved in the feature branch before final merge

File: tests/integration/test_kfp_functional_v1.py
Comment (orfeas-k) [inline tests/integration/test_kfp_functional_v1.py:112]: Regarding this, I had changed this to `30` [here](https://github.com/canonical/kfp-operators/pull/350/files#diff-ad79bdbee1716da8d0eebee39133a7e79767b69755aef15512553ad4a1588ad8L77-R77) since this made the CI flaky and assertion was failing although the charms were all active. I remember having to run the tests many times (both in the CI and locally) in order for them to succeed.

File: tests/integration/test_kfp_functional_v1.py
Comment (orfeas-k) [inline tests/integration/test_kfp_functional_v1.py:40]: You could use `KFP_SDK_VERSION` here, right?
```suggestion
        uploadfile=SAMPLE_PIPELINE[KFP_SDK_VERSION], name=SAMPLE_PIPELINE_NAME
```

File: tests/integration/test_kfp_functional_v2.py
Comment (orfeas-k) [inline tests/integration/test_kfp_functional_v2.py:39]: Same here, you could use `KFP_SDK_VERSION` here, right?
```suggestion
        uploadfile=SAMPLE_PIPELINE[KFP_SDK_VERSION], name=SAMPLE_PIPELINE_NAME
```

File: tests/integration/test_kfp_functional_v2.py
Comment (orfeas-k) [inline tests/integration/test_kfp_functional_v2.py:113]: Here we have 30 seconds. WRT my other comment, if there isn't a reason to have different idle_periods when using sdk_v1 and sdk_v2, we should decide and keep those the same in both places.

File: tests/integration/test_kfp_functional_v1.py
Comment (orfeas-k) [reply tests/integration/test_kfp_functional_v1.py:112]: Just to clarify, they were failing because one of the charms would go to `Maintenance` status during this long `idle_period`, and the assertion would end up timing out.

File: tests/integration/test_kfp_functional_v2.py
Comment (orfeas-k) [inline tests/integration/test_kfp_functional_v2.py:174]: nit but should we name this `recurring-job-KFP_SDK_VERSION` since we 've also did
```
    pipeline_name = f"test-pipeline-sdk-{KFP_SDK_VERSION}"
...
        run_name=f"test-run-sdk-{KFP_SDK_VERSION}",
```

If yes, we could do that for v1 too.

**Review by orfeas-k — COMMENTED** [review]
Thank you @DnPlas for the great work on KFP. I left some minor comments.

File: tests/integration/test_kfp_functional_v1.py
Comment (DnPlas) [reply tests/integration/test_kfp_functional_v1.py:112]: It was not failing for me, in fact the bundle tests ran and succeeded at the first time.

File: tests/integration/test_kfp_functional_v2.py
Comment (DnPlas) [reply tests/integration/test_kfp_functional_v2.py:113]: I don't think this was intentional, maybe a rebase error. I think I can leave it on 30s, let me just change it to 30s based also in your other comment.

File: tests/integration/test_kfp_functional_v2.py
Comment (DnPlas) [reply tests/integration/test_kfp_functional_v2.py:174]: Sure, we could do that.

**Pull request merged**
